### ISSUE TITLE
Add Beni Tracker module support

### ISSUE THREAD

**Issue by Sembiance** [libxmp/libxmp#460]
Would be awesome if libxmp supported Beni Tracker. Info, file samples and more: http://fileformats.archiveteam.org/wiki/Beni_Tracker_module

I haven't been able to locate ANY modern player. The wiki link above has a link to adplug which seems to have some sample source code for playing the files, but doesn't actually appear to work with AdPlug.

Format info:
```
The file extension .PIS is used since the previous incarnation
of Beni Tracker was called "Psychisch Tracker" (unreleased).

Offset   Bytes    Description
------   -----    -----------
0        1        Song length
1        1        N. of patterns
2        1        N. of instruments
3        ?        Pattern map (N. of patterns * 1 byte)
?        ?        Instrument map (N. of instruments * 1 byte)
?        ?        Order list (Song length * 9 bytes)
?        ?        Pattern data (N. of patterns * 192 bytes)
?        ?        Instrument data (N. of instruments * 11 bytes)
?        4        "B.J." (v1.8 identification mark)

-Pattern and instrument maps
   PIS, the intelligent format it is, saves only the
   patterns and instruments that are used. The maps
   therefor contain for each pattern (instrument) in
   the file the number that was used for it during editing.
   Eg. if You have this in Your (otherwise empty) order list:
      02
      04
      00
   The pattern map, containing the indices of used patterns
   in ascending order, will look like this:
      ptnmap[0]=0
      ptnmap[1]=2
      ptnmap[2]=4
   Ergo, the maps contain the order in that the patterns
   (instruments) are saved in the data blocks below.

-Pattern data format
   A pattern consists of 64 * 3 bytes where 3 bytes encode
   one note as:

      nnnn.... ........ ........
      ....ooo. ........ ........
      .......i iiii.... ........
      ........ ....eeee eeeeeeee

   n = note (0..12) as index into frequency table, 12 is blank
   o = octave
   i = instrument
   e = effect

-Instrument format
   An instrument consists of 11 bytes.

   Offset   Description
   ------   -----------
   0        Op.1: Multiple/EG type/Vibrato/Amp.mod.
   1        Op.2:                 "
   2        Op.1: Output level
   3        Op.2:    "     "
   4        Op.1: Attack/Decay
   5        Op.2:       "
   6        Op.1: Sustain/Release
   7        Op.2:        "
   8        Op.1: Waveform
   9        Op.2:     "
   10       Feedback/Connection type

-"B.J."
   To let v1.8 distinguish between module formats. Has absolutely
   no effect when loaded in v1.3.
```

**Comment by AliceLR** [comment]
libxmp tried to support AdLib formats in the past but allegedly the support was very bad, hence it was removed. This looks like something that is more suited to AdPlug. (From a glance I don't even see a loader for this in AdPlug's source tree, but I'm not very familiar with how it's organized.)

**Comment by sagamusix** [comment]
Indeed, this format doesn't belong into libxmp at least at this point in time as OPL support was removed. I started writing a loader for OpenMPT and it mostly works, apart from pitch slides (which are very different from other formats). Strangely when running the tracker in DOSBox all instruments use the default Sine waveform even though the instruments are configured to use different instruments. Looking at the replayer code, I see nothing suspicious so I'll have to double-check on real hardware if this is a player bug or a DOSBox bug.

**Comment by AliceLR** [comment]
Thoughts/potential "roadmap" about how to go forward with this. If this happens, this will probably all end up being on me, and it would only happen after I finish a few fixes for IT...:

1) Replace `fmopl.c` with a permissive licensed C OPL3 emulator. Opal is PD but is C++ and has bugs that annoyingly haven't had their fixes released officially by Reality. The DOSBox emulators IIRC are LGPL or GPL and also have bugs. Nuked is LGPL 2.1. ymfm is BSD 3-clause and C++. The easiest path forward here might be a C port of Opal with the bugfixes applied and its terrible/slow linear resampler removed (annoying, but easy).
2) (Re)implement software mixer support for OPL3 emulator (easy).
3) (Re)implement OPL support for S3M to iron out the effects transforms/playercode port calls (moderate to difficult).
4) Implement a Beni Tracker loader (format is simple, so the difficulty depends on the effects...).

This format doesn't use MIDI pattern data so at least it has that going for it (CMF... 🙄).

As a side note, I think libxmp's existing extras infrastructure paired with fixed OPL support *might* be adequate for RAD v2. Which, "unrelatedly", desperately needs a tracker that isn't effectively abandoned with crash bugs and an OPL emulator with backwards key scale level. I assume I'll have to do that too. :(

**Comment by Wohlstand** [comment]
Little note for Opal emulator lacks the rhythm mode at all, so, need to be completely implemented to support it. What about Nuked OPL3 emulator, it supports the chip very accurately, it's the most accurate chip emulator I know, and it's heavy in comparison to others because of its accuracy. Also, let's ask @nukeykt (an author of Nuked OPL3 emulator), what he thinks about the whole thing?

**Comment by AliceLR** [comment]
> Little note for Opal emulator lacks the rhythm mode at all

> What about Nuked OPL3 emulator,

See issue #387, both are addressed there.

**Comment by sagamusix** [comment]
> Little note for Opal emulator lacks the rhythm mode at all, so, need to be completely implemented to support it.

Which trackers support rhythm mode at all? It seems very painful to use in classic tracker setup, and I haven't seen it anywhere (which is one reason why OpenMPT uses Opal, the license being another).

**Comment by AliceLR** [comment]
> Which trackers support rhythm mode at all?

It's not technically a "tracker" format, but Creative Music Format (.CMF) is the main one I'm aware of.

**Comment by sagamusix** [comment]
Yeah, as that is essentially a MIDI file with OPL instruments I'd say that adding support for that is a big nope if/when OPL support gets re-introduced.
